sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+of `__devicelib_assert_fail` (i.e. if Device-side Runtime doesn't support it).
+
+Any properties in `isAssertEnabled` property set are ignored.
+
+The property set is added to device binary descriptor whenever at least single
+translation unit was compiled with assertions enabled i.e. `NDEBUG` undefined.
+
+The property set is added by `sycl-post-link` tool depending on module metadata.
+Metadata is provided by Clang frontend. Metadata name is `is_assert_enabled`.
+
+
+##### Compiling with assert enabled/disabled
+
+Consider the following example sources:
+```c++
+// impl.cpp
+using namespace sycl;
+int calculus(int X) {
+  assert(X && "Invalid value");
+  return X * 2;
+}
+
+void enqueueKernel(queue &Q, buffer &B) {
+  Q.submit([](handler &H) {
+    auto Acc = B.get_access<read_write>(H);
+    H.parallel_for(/* range */, [](item It) {
+      assert(Acc[It]);
+      // ...
+    });
+  });
+}
+
+// main.cpp
+// ...
+using namespace sycl;
+
+SYCL_EXTERNAL int calculus(int);
+void enqueueKernel(queue&, buffer&);
+
+void workload() {
+  queue Q;
+  buffer B;
+
+  Q.submit([](handler &H) {
+    auto Acc = B.get_access<read_write>(H);
+    H.parallel_for(/* range */, [](item It) {
+      int X = calculus(0); // should fail assertion
+      assert(X && "Nil in result");
+      Acc[It] = X;
+    });
+  });
+
+  enqueueKernel(Q, B);
+  ...
+}
+```
+
+These two files are compiled into a single binary application. There are four
+states of definedness of `NDEBUG` macro available:
+
+| # | `impl.cpp` | `main.cpp` |
+| - | ---------- | ---------- |
+| 1 | defined    | defined    |
+| 2 | defined    | undefined  |
+| 3 | undefined  | defined    |
+| 4 | undefined  | undefined  |
+
+States of definedness of `NDEBUG` macro defines the set of assertions which can
+fail. Having assertions enabled in at least one translation unit with device
+code requires for `isAssertEnabled` property set being present in device image
+descriptor structure.
+
+
+### Storing accessor metadata and writing assert failure to buffer
+
+Both storing of accessor metadata and writing assert failure is performed with
+help of built-ins. Implementations of these builtins are substituted by
+frontend.
+
+User's kernel is executed through a wrapper. Wrapping takes place in DPCPP
+Runtime headers in a following manner:
+
+```
+class handler {
+
+template <typename KernelName> parallel_for(KernelFunc, Range) {
+#ifndef NDEBUG
+  // Assert required
+  if (!MQueue->get_device()->assert_fail_supported()) {
+    using KName2 = class ASSERT_WRAPPER_NAME(KernelName);
+    
+    auto AssertBufferAcc = MQueue->get_context()->getAssertBufferAccessor(this);
+
+    parallel_for_impl<KName2>(
+      Range,
+      [=](Item) {
+        __store_acc(AssertBuffAcc);
+        KernelFunc(Item);
+      });
+  } else {
+#endif
+
+     // (No assert required) OR (Assert supported by device)
+     // ordinary enqueue process
+
+#ifndef NDEBUG
```

Comment:
This effectively duplicates amount of kernels if assertions are enabled: even if device supports them natively, the device compiler still sees and emits the second version of kernel with fallback assert implementation. In combination with range rounding feature this increases original amount of kernels by 4x, which doesn't look right.

Would it be better to avoid kernels duplication here and just always instrument them with assert buffer regardless of native support for the feature? This will just increase arguments size by 24 bytes (If I recall correctly how huge accessor is) and we can also try and tune our dead argument elimination optimization to remove it completely.

Reply:
This duplication would be absolutely devastating on FPGA where each additional copy of a kernel costs area on the chip even if it is never used.  We disable the range rounding feature on FPGA for precisely this reason.  

To specialize the kernel body code (but not the kernel arguments) you could use: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceIf/device_if.asciidoc.  The code would look something like this:

```c++
  kh.ext_oneapi_if_device_has<aspect::assert>([&]() {
    __store_acc(AssertBuffAcc);
  });
```

where kh is a kernel_handler object passed to the kernel lambda.  Since you'd need to also wrap your __store_assert_failure() in the same way you'd need to save the kernel_handler object in a program-scope global as well so that you don't have to pass it along to each function call.

Reply:
As an option suggest avoiding changing original users kernels and instead introduce a helper kernel which sets global buffer to desired value. This kernel helper would need to be executed as initialization of a device program before any other kernel(aka global constructor).